### PR TITLE
perf: wait-free WAL

### PR DESCRIPTION
**Pull request by CJP10**
https://github.com/influxdata/influxdb_iox/issues/437

### CODE REVIEW

**Comment by e-dard** [comment]
Hey @CJP10, thanks for the contribution.

I can't see your acceptance of our CLA. Can you sign it if you haven't yet? The form is linked from the README on the project.

**Review by alamb — COMMENTED** [review]
The use of `crossbeam_epoch` is cool -- I learn something cool every day. 

In general, I guess I am wondering "how did you determine that this code was faster than the existing code"?  I also wonder if you could share any additional thoughts about the design at a high level or perhaps add some tests showing the intent of the design (e.g. show what situation `using_last_active` is for.


Also, I checked this branch out locally and there seem to be several compile errors

```
alamb@MacBook-Pro influxdb_iox2 % cargo test --all
cargo test --all
   Compiling wal v0.1.0 (/Users/alamb/Software/influxdb_iox2/wal)
   Compiling criterion v0.3.3
   Compiling cluster v0.1.0 (/Users/alamb/Software/influxdb_iox2/cluster)
   Compiling influxdb_iox v0.1.0 (/Users/alamb/Software/influxdb_iox2)
error[E0599]: no method named `total_size` found for struct `wal::Wal` in the current scope
  --> wal/tests/total_size.rs:22:20
   |
22 |     assert_eq!(wal.total_size(), 0);
   |                    ^^^^^^^^^^ private field, not a method

error[E0308]: mismatched types
  --> wal/tests/no_concurrency.rs:14:38
   |
14 |     let sequence_number = wal.append(payload)?;
   |                                      ^^^^^^^ expected struct `std::vec::Vec`, found struct `wal::WritePayload`
   |
   = note: expected struct `std::vec::Vec<u8>`
              found struct `wal::WritePayload`

error[E0308]: mismatched types
  --> wal/tests/helpers/mod.rs:70:25
   |
70 |               $wal.append(data)?;
   |                           ^^^^ expected struct `std::vec::Vec`, found struct `wal::WritePayload`
   | 
  ::: wal/tests/delete-up-to.rs:16:5
...
```

So all in all I am not sure how to proceed with this PR

**Comment by CJP10** [comment]
@alamb Sorry I didn't make it clear, this is still quite WIP. I'm not sure what y'alls convention is for WIP but this would definitely qualify as WIP.

 I'll fix up the tests whens it more feature complete. I'll also do my best to get [loom](https://github.com/tokio-rs/loom) working to verify some concurrency edge cases.

In terms of hard performance numbers, I don't have those. The impl isn't done yet so I'd have to cross that bridge first. I'd also have to write some benchmarks for both the new impl and old impl. I didn't see any benches for the old impl but I could have missed them. I'm currently operating under  some assumptions/educated guesses. One of those being that using the parallelism of the entire runtime and not using channels to middleman the syscalls will be a positive impact on performance.

I guess I can go over a bit of the impl and hope it does justice or sparks some more questions.

The reason for using `crossbeam_epoch` to wrap the WAL files is to avoid a use-after-free post rollover.
Drilling down a bit, when threads call append and pass the `rollover_size` they race to mark the `active_file` ptr.
The thread that marks the pointer is responsible for creating the new WAL file and CAS the `active_file` with the one created.
`last_active_file` exists because a task can be preempted during a rollover and come back to execute after the rollover.
The thread will then use `last_active_file` to write that payload. This isn't perfect because the task can be preempted indefinitely and so the correct approach is to store a list of every WAL file so you write back to any time. `ROLLOVER_SEQUENCE_NUMBER_BUFFER` also plays a part as it allows further writing to a WAL file after rollover. The current value is somewhat arbitrary but finding a sensible default seems straight forward. Regardless, this is needed because the exact rollover point can't be determined so we leave room for outstanding writes to the previous WAL file.

Int terms of safety, a majority of the unsafe is documented on the crossbeam_epoch side but I wouldn't mind putting a comment block on each giving some more detail.

File: wal/src/lib.rs [hunk 1]
```diff
@@ -200,139 +213,272 @@ impl WalBuilder {
...
+        let mut active_file = unsafe {
+            self.active_file
+                .load(Acquire, &guard)
+                .as_ref()
+                .expect("active_file ptr was null")
         };
 
-        let h = Header {
+        if sequence_number < active_file.start_sequence_number {
+            unsafe {
```
[the remark this answers is not in the document]
Comment (CJP10) [reply wal/src/lib.rs:280]: Checkout my other comments if this piece is still unclear I'm happy to dig in more.

**Comment by e-dard** [comment]
@CJP10 I'm going to work on some more contribution documentation next week. I don't know why Github hasn't enabled Draft PRs on this repo but for now I have prefixed your title with `[WIP]` and added a label.

**Comment by alamb** [comment]
> Sorry I didn't make it clear, this is still quite WIP. I'm not sure what y'alls convention is for WIP but this would definitely qualify as WIP.

@CJP10  Cool -- that makes sense -- sorry I was confused. 

Also, thank you for the explanation

**Comment by CJP10** [comment]
@alamb No worries! I'll refine the impl a bit more and when it's ready I'll shoot you a ping if that's cool.

**Comment by e-dard** [comment]
Confirmation @CJP10 has signed the CLA.

**Comment by CJP10** [comment]
@alamb there is still some plumbing left to do but the overall impl of `append` is pretty much done. If you could give it a read and let me know what you think?

**Review by alamb — COMMENTED** [review]
Thanks again for this work @CJP10

As @pauldix mentioned on https://github.com/influxdata/influxdb_iox/issues/437#issuecomment-725392018 , we don't really evision the performance of the local file system-based WAL will be a focus of this project. So I don't want to lead you on working on this PR unnecessarily. 

I took a quick read through some of this PR

In terms of the design, I wonder if you have any benchmarks or usecases that were driving this implementation? As in, I am sure this design is better is several aspects than the current code, but it does have drawbacks too (like the use of `unsafe`)?

Also, can you explain why you removed the existing `wal/tests` test files? The scenarios in them seem to still be relevant to this implementation. Also I didn't see many (any?) tests for the new code that was added. 

Also, can you explain what guarantees this implementation gives compared to the current WAL (which upon completion of `write_and_sync` guarantees the data has been written to the wal and `fsync` has returned, giving some additional durability to the data).  I probably don't understand fully, but it seems like this implementation purposely doesn't wait for fsync, which probably does increase performance but at the expense of durability.

**Comment by CJP10** [comment]
> In terms of the design, I wonder if you have any benchmarks or usecases that were driving this implementation? As in, I am sure this design is better is several aspects than the current code, but it does have drawbacks too (like the use of `unsafe`)?

On my machine (Ryzen 5950x@Stock and Samsung 980 Pro 500GB) I was able to get ~4700MB/s (~5000MB/s is what I got just using fio to benchmark the disk). So it's pretty close to running at raw disk speed without fsyncing. I don't think you can get match faster than the current append hot path. I came up with this when implementing a K-V store based on [WiscKey](https://www.usenix.org/system/files/conference/fast16/fast16-papers-lu.pdf). Unfortunately, `unsafe` is the nature of the beast and can't really be avoided in this impl.

> Also, can you explain why you removed the existing `wal/tests` test files? The scenarios in them seem to still be relevant to this implementation. Also I didn't see many (any?) tests for the new code that was added.

I took some inspirations from the existing tests but there wasn't much carry over. The file rollover test didn't apply because file rollover isn't deterministic and so verifying the exact details of the files wasn't of value. The delete up to test wouldn't provide much value because it's just deleting a file. I wouldn't mind adding back the old tests but the new tests do cover most of the edge cases (located at the bottom of the lib.rs).


> Also, can you explain what guarantees this implementation gives compared to the current WAL (which upon completion of `write_and_sync` guarantees the data has been written to the wal and `fsync` has returned, giving some additional durability to the data). I probably don't understand fully, but it seems like this implementation purposely doesn't wait for fsync, which probably does increase performance but at the expense of durability.

Added syncing :smile:.

**Comment by alamb** [comment]
Thanks @CJP10 

> On my machine (Ryzen 5950x@Stock and Samsung 980 Pro 500GB) I was able to get ~4700MB/s (~5000MB/s is what I got just using fio to benchmark the disk

Is it possible to share the benchmark / scripts you were using? Did you compare this PR's performance to the performance on `main`? 

The comparison that is probably the most relevant is when `fsync` is enabled

**Comment by alamb** [comment]
The other thing I wonder as I spend time reading about crossbeam/epoch is if you tested this implementation under concurrent load (aka when there are  multiple writers across different threads), aka in the type of scenario that I would expect unsafe behavior to potentially be tricky to get right

Our project does not yet have this type of tooling / benchmark, so I don't have any specific suggested commands

File: wal/src/io.rs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    header_bytes: &[u8],
+    data_bytes: &[u8],
+    offset: u64,
+) -> Result<(), IoError> {
+    let iovec = [
+        IoVec::from_slice(header_bytes),
+        IoVec::from_slice(data_bytes),
+    ];
+
+    pwritev(file.as_raw_fd(), &iovec, offset as i64).context(FailedToWriteDataUnix)?;
```
Comment (alamb) [inline wal/src/io.rs:33]: I apologize for my systems ignorance, but I don't understand why it is safe to concurrently write to a file using `pwritev` but not with `file.write_all`.

File: wal/src/payload.rs [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    }
+
+    pub fn as_bytes(&self) -> [u8; Self::LEN] {
+        let mut bytes = [0u8; Self::LEN];
+        bytes[0..1].copy_from_slice(&self.magic.to_le_bytes());
+        bytes[1..5].copy_from_slice(&self.checksum.to_le_bytes());
+        bytes[5..9].copy_from_slice(&self.len.to_le_bytes());
+        bytes
+    }
+}
```
Comment (alamb) [inline wal/src/payload.rs:171]: Some tests for the encoding / decoding round trip might be good here -- bonus points for adding coverage for the error cases

File: wal/src/lib.rs [hunk 4]
```diff
@@ -6,738 +6,582 @@
...
-            r => r
-                .map(Some)
-                .context(UnableToOpenFile { path })
-                .map_err(Into::into),
+    pub fn append(&self, data: &[u8]) -> Result<SequenceNumber> {
+        let guard = epoch::pin();
+        // SAFETY: file is properly synchronized (Acquire ordering)
+        // so we won't be dereferencing uninitialized data
+        let active_wal = self.active.load(Acquire, &guard);
+        let active_wal_ref = unsafe { active_wal.as_ref() }.expect("active file was null!");
```
Comment (alamb) [inline wal/src/lib.rs:280]: Is it correct that the core of the idea is here: namely that this unsafe code allows multiple threads to write concurrently to the same file descriptor?

I probably misunderstand the design, but if that is indeed the case, couldn't we just copy the file descriptor?

In general, I don't understand is how this code ensures the WAL data isn't corrupted by multiple concurrent writers

**Review by alamb — COMMENTED** [review]
In general, I don't understand is how this code ensures the WAL data isn't corrupted by multiple concurrent writers. I probably don't understand something about the design (see comment on `append`)

I was trying to play with this code locally, but it didn't compile for me (on my MacOS laptop):

```
   Compiling generated_types v0.1.0 (/Users/alamb/Software/influxdb_iox2/generated_types)
error[E0432]: unresolved import `nix::sys::uio::pwritev`
 --> wal/src/io.rs:1:21
  |
1 | use nix::sys::uio::{pwritev, IoVec};
  |                     ^^^^^^^
  |                     |
  |                     no `pwritev` in `sys::uio`
  |                     help: a similar name exists in the module: `writev`

error[E0433]: failed to resolve: use of undeclared type `SeekFrom`
  --> wal/src/io.rs:56:15
   |
56 |     file.seek(SeekFrom::Start(offset)).context(FailedToSeek)?;
   |               ^^^^^^^^ not found in this scope
   |
```

File: wal/src/payload.rs (hunk 3 above)
Comment (CJP10) [reply wal/src/payload.rs:171]: Agreed

File: wal/src/io.rs (hunk 2 above)
Comment (CJP10) [reply wal/src/io.rs:33]: Give the [man page](https://linux.die.net/man/2/pwritev) a read, it's pretty helpful. To give a summary pwritev is atomic and that's handled at the OS level. The std-lib has a pwrite impl [write_all_at](https://doc.rust-lang.org/std/os/unix/fs/trait.FileExt.html#method.write_all_at) which only require a `&File`. `pwritev` just allows vectored writes which prevent a realloc in user land.

File: wal/src/lib.rs (hunk 4 above)
Comment (CJP10) [reply wal/src/lib.rs:280]: You can think of `self.active` as just a `&File`. The reason it's not just a simple `&File` is due to the requirement of replacing the active file during rollover. So you could just use a wrap it in a `RwLock` and call it a day. The downside being that rolling over a file (calling write() on the RwLock to get mutable access) can block indefinitely. `RwLock` will kill latencies on the tail due to the rollover blocking all read(appending only needs a `&File`) to `self.active` .




> In general, I don't understand is how this code ensures the WAL data isn't corrupted by multiple concurrent writers. I probably don't understand something about the design (see comment on `append`)
> 
> I was trying to play with this code locally, but it didn't compile for me (on my MacOS laptop):
> 
> ```
>    Compiling generated_types v0.1.0 (/Users/alamb/Software/influxdb_iox2/generated_types)
> error[E0432]: unresolved import `nix::sys::uio::pwritev`
>  --> wal/src/io.rs:1:21
>   |
> 1 | use nix::sys::uio::{pwritev, IoVec};
>   |                     ^^^^^^^
>   |                     |
>   |                     no `pwritev` in `sys::uio`
>   |                     help: a similar name exists in the module: `writev`
> 
> error[E0433]: failed to resolve: use of undeclared type `SeekFrom`
>   --> wal/src/io.rs:56:15
>    |
> 56 |     file.seek(SeekFrom::Start(offset)).context(FailedToSeek)?;
>    |               ^^^^^^^^ not found in this scope
>    |
> ```

Yeah that's my bad I didn't had a non-unix machine to test on hand. I'll have to cfg the `nix` dep and some other deps.

**Comment by CJP10** [comment]
> Thanks @CJP10
> 
> > On my machine (Ryzen 5950x@Stock and Samsung 980 Pro 500GB) I was able to get ~4700MB/s (~5000MB/s is what I got just using fio to benchmark the disk
> 
> Is it possible to share the benchmark / scripts you were using? Did you compare this PR's performance to the performance on `main`?
> 
> The comparison that is probably the most relevant is when `fsync` is enabled

I didn't compare perf but I can commit my benches and port them to main for the existing WAL.

> In general, I don't understand is how this code ensures the WAL data isn't corrupted by multiple concurrent writers. I probably don't understand something about the design (see comment on `append`)
> 
> I was trying to play with this code locally, but it didn't compile for me (on my MacOS laptop):
> 
> ```
>    Compiling generated_types v0.1.0 (/Users/alamb/Software/influxdb_iox2/generated_types)
> error[E0432]: unresolved import `nix::sys::uio::pwritev`
>  --> wal/src/io.rs:1:21
>   |
> 1 | use nix::sys::uio::{pwritev, IoVec};
>   |                     ^^^^^^^
>   |                     |
>   |                     no `pwritev` in `sys::uio`
>   |                     help: a similar name exists in the module: `writev`
> 
> error[E0433]: failed to resolve: use of undeclared type `SeekFrom`
>   --> wal/src/io.rs:56:15
>    |
> 56 |     file.seek(SeekFrom::Start(offset)).context(FailedToSeek)?;
>    |               ^^^^^^^^ not found in this scope
>    |
> ```

That's my bad I have to cfg that dep anyway on non-supported platforms.

File: wal/src/lib.rs (hunk 4 above)
Comment (CJP10) [reply wal/src/lib.rs:280]: `self.active` is basically a `&FIle` in its use. File is just a fd with drop magic to close it. The issue with just copying around a fd is that tracking the usage becomes difficult when many threads are using the same fd corruently. `crossbeam-epoch` basically just does the gc of the file once all refs to it are gone. It's similar to an Arc however it lets the value be swapped. The `arcswap` crate could also be using here instead of `crossbeam-epoch`.

There a two invariants that need to hold, 1) we don't use a FIle after it's dropped 2) we don't have overlapping writes.
1) is handled by `crossbeam-epoch`
2) is handled by the Atomicu64 on the file's size field. `pwritev` also ensures that this down to the OS layer.

**Comment by alamb** [comment]
@CJP10  -- I have spent more time playing around with this code and I think I am finally beginning to grok what is going on (namely that multiple threads don't clobber each other's writes because the offset/seek of the file is calculated in `append`). It is pretty clever.

My current feeling is that
1. If we can show significant performance improvements of this implementation over the current implementation on main we could move to the approach shown in this PR.
2. I would like to avoid the use of `unsafe` if at all possible, and so I would likely spend time trying to avoid its use here (I am very curious how much extra performance the use of `unsafe` adds, as mentioned below). 

> I didn't compare perf but I can commit my benches and port them to main for the existing WAL.

If you can simply share how you are testing this WAL implementation (how you measured the numbers you reported above) I can take a shot at porting it back to master and getting some numbers.


For what it is worth,  I wrote a multi-threaded test (https://github.com/influxdata/influxdb_iox/commit/01572c0f83e4cb8fd265a3f666a9030d1de17ed2) to try to trigger some incorrect results but was not successful. :+1: to the code. 

Also, something I noticed on my (very under powered) linux machine is that the `wal_iter` test is very slow (37 seconds) and it is ~ entirely IO bound (which isn't surprising I suppose, since every call to `append` results in a system call and given the crappy disk in my laptop). However but it makes me wonder how much difference the `unsafe` code makes for performance. 

Here is the data in case that is interesting

```
andrew@trogdor:~/Software/influxdb_iox/wal$ time cargo test --release -- wal_iter
...
    Finished release [optimized + debuginfo] target(s) in 0.14s
     Running /home/andrew/Software/influxdb_iox/target/release/deps/wal-cfbb1f964b48714a

running 1 test
test tests::wal_iter ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 5 filtered out

real	0m37.445s
user	0m0.344s
sys	0m0.221s
```

And looking at top that was all waiting on IO

File: wal/src/lib.rs (hunk 4 above)
Comment (alamb) [reply wal/src/lib.rs:280]: After more messing around, with respect to 1,  I still don't understand how we could have some thread writing to the File and then have the File dropped. It seems like the only way the File can be closed is if the Wal itself is dropped, but there is Rust code somewhere that is calling `Wal::append` and thus `self` will be valid

I would probably have to try and change the code to avoid crossbeam-epoch to see what is wrong with my idea. That is something I will once I understand the performance improvement that this approach can achieve.

Comment (CJP10) [reply wal/src/lib.rs:280]: @alamb The issue with `Arc` is swapping the value on rollover. You'd need to have something like `RwLock<Arc<File>>`.

Comment (alamb) [reply wal/src/lib.rs:280]: Right -- exactly. Which would likely have a bit of CPU overhead for sure

**Comment by alamb** [comment]
#TLDR
It is my opinion this PR doesn't improve performance sufficiently to justify the additional complexity introduced (in terms of using `unsafe`, having os specific compilation flags, and additional dependencies).

Maybe I am missing something, but I don't think we should take this change given what I have seen.

I would be happy to contemplate a WAL change that reduced the calls to write using safe code. @CJP10  we sort of failed on this PR as we don't yet have clear contribution guidelines here that would have made the decision making process more transparent and could have focused us more on performance benchmarks)


# Discussion:
I compared the performance of this branch against the WAL on master using the wal bench provided (along with variants that wrote the same data 10 and 100 times)

While I saw a 30% improvement on this branch compared to master, I believe almost all of that is attributable to using fewer system IO calls on each write (the current WAL implementation on main does a few using unbuffered writes). When I simply added buffering into the WAL on main it goes  5 times faster than either implementation.

Not that I am advocating using buffered IO in the WAL, but I am using that observation as additional evidence that the WAL is (unsurprisingly) entirely IO bound; Allowing multiple threads to write concurrently (as in this PR) does not change the fact it is IO bound nor change its performance characteristics.

Here are some performance results using  my macbook pro

# This PR

Using the WAL code in this PR + some fixes to get it to compile on mac: https://github.com/influxdata/influxdb_iox/tree/alamb/wal-with-bench @ a91e96ea1fcc28335ea0852e9ff8c4d5536163dd

```

wal/append-1            time:   [8.9782 us 9.0758 us 9.2008 us]
                        thrpt:  [424.55 MiB/s 430.40 MiB/s 435.08 MiB/s]


wal/append-10           time:   [89.346 us 90.176 us 91.093 us]
                        thrpt:  [428.82 MiB/s 433.18 MiB/s 437.20 MiB/s]


wal/append-100          time:   [880.35 us 887.56 us 896.55 us]
                        thrpt:  [435.70 MiB/s 440.11 MiB/s 443.72 MiB/s]
````

# Main

Against the WAL on main (without calls to `sync_all` so I think it is a reasonably fair comparison): https://github.com/influxdata/influxdb_iox/tree/alamb/wal-perf-on-main @ 0999adbe2663bd9d4704b6ec775db2e0e54b3364

```
wal/append-1            time:   [10.649 us 10.736 us 10.827 us]
                        thrpt:  [360.79 MiB/s 363.85 MiB/s 366.80 MiB/s]

wal/append-10           time:   [107.82 us 110.23 us 113.80 us]
                        thrpt:  [343.27 MiB/s 354.36 MiB/s 362.30 MiB/s]

wal/append-100          time:   [1.0710 ms 1.0795 ms 1.0884 ms]
                        thrpt:  [358.91 MiB/s 361.85 MiB/s 364.73 MiB/s]
```


# Main using buffered IO

When using a buffered writer (which isn't actually a good idea, but since we aren't sync'ig the writes anyways, I don't think it changes the durability story materially):

On https://github.com/influxdata/influxdb_iox/tree/alamb/wal-with-buffering at 4af4ac67ba0df90d61f2411ddd6f23a09c15d309


```
wal/append-1            time:   [1.6421 us 1.6674 us 1.7033 us]
                        thrpt:  [2.2396 GiB/s 2.2878 GiB/s 2.3231 GiB/s]

wal/append-10           time:   [16.681 us 17.064 us 17.731 us]
                        thrpt:  [2.1514 GiB/s 2.2355 GiB/s 2.2868 GiB/s]

wal/append-100          time:   [167.25 us 168.83 us 170.52 us]
                        thrpt:  [2.2371 GiB/s 2.2595 GiB/s 2.2808 GiB/s]
```

cc @CJP10

**Comment by alamb** [comment]
Here are the measurements from my (old crappy) linux laptop. While the magnitudes differ, the overall pattern is the same


https://github.com/influxdata/influxdb_iox/tree/alamb/wal-with-bench (this PR):
```
wal/append-1            time:   [5.2238 us 5.2879 us 5.3651 us]
                        thrpt:  [728.09 MiB/s 738.72 MiB/s 747.77 MiB/s]

wal/append-10           time:   [53.051 us 53.299 us 53.580 us]
                        thrpt:  [729.05 MiB/s 732.90 MiB/s 736.32 MiB/s]

wal/append-100          time:   [531.34 us 535.37 us 540.01 us]
                        thrpt:  [723.36 MiB/s 729.64 MiB/s 735.17 MiB/s]
```

https://github.com/influxdata/influxdb_iox/tree/alamb/wal-perf-on-main

```                                                                              
wal/append-1            time:   [9.4452 us 9.4983 us 9.5605 us]
                        thrpt:  [408.58 MiB/s 411.26 MiB/s 413.57 MiB/s]

 wal/append-10           time:   [98.074 us 98.481 us 98.938 us]
                        thrpt:  [394.82 MiB/s 396.65 MiB/s 398.30 MiB/s]

wal/append-100          time:   [988.17 us 996.37 us 1.0065 ms]
                        thrpt:  [388.10 MiB/s 392.05 MiB/s 395.30 MiB/s]
```

https://github.com/influxdata/influxdb_iox/tree/alamb/wal-with-buffering

```
wal/append-1            time:   [2.6238 us 2.6593 us 2.6991 us]
                        thrpt:  [1.4133 GiB/s 1.4345 GiB/s 1.4539 GiB/s]

wal/append-10           time:   [26.494 us 26.791 us 27.095 us]
                        thrpt:  [1.4079 GiB/s 1.4239 GiB/s 1.4399 GiB/s]

wal/append-100          time:   [262.86 us 264.67 us 266.52 us]                           
                        thrpt:  [1.4313 GiB/s 1.4413 GiB/s 1.4512 GiB/s]

```
